modules/lnd/lnd_wrapper/wrapper.go

```diff
@@ -184,6 +191,145 @@ func LndParseP2pLightningMessage(data *C.char, length C.int) *C.char {
 		sb.WriteString(fmt.Sprintf("%d", fc.FundingPoint.Index))
 		sb.WriteString(";SIGNATURE=")
 		sb.WriteString(fmt.Sprintf("%x", fc.CommitSig.ToSignatureBytes()))
+	case 38:
+		messageShutdown := message.(*lnwire.Shutdown)
+		if messageShutdown.ExtraData != nil || messageShutdown.CustomRecords != nil || messageShutdown.ShutdownNonce.IsSome() {
+			return nil
+		}
+		sb.WriteString("MSG_TYPE=shutdown;CHANNEL_ID=")
+		sb.WriteString(fmt.Sprintf("%x", messageShutdown.ChannelID[:]))
+		sb.WriteString(";SCRIPTPUBKEY=")
+		sb.WriteString(fmt.Sprintf("%x", messageShutdown.Address))
+	case 39:
+		messageClosingSigned := message.(*lnwire.ClosingSigned)
+
+		// LND actually doesn't check the signature when parsing the message,
+		// but we do, otherwise the fuzzer will crash all the time.
+		_, err := messageClosingSigned.Signature.ToSignature()
+		if err != nil {
+			return C.CString("")
+		}
+		// FeeSatoshis should be u64 but it's i64 in LND
+		if messageClosingSigned.FeeSatoshis < 0 {
+			return nil
+		}
+		// If there is any extra data skip.
+		if messageClosingSigned.ExtraData != nil {
+			return nil
+		}
+		sb.WriteString("MSG_TYPE=closing_signed;CHANNEL_ID=")
+		sb.WriteString(fmt.Sprintf("%x", messageClosingSigned.ChannelID[:]))
+		sb.WriteString(";FEE_SATOSHIS=")
+		sb.WriteString(fmt.Sprintf("%d", messageClosingSigned.FeeSatoshis))
+		sb.WriteString(";SIGNATURE=")
+		sb.WriteString(fmt.Sprintf("%x", messageClosingSigned.Signature.ToSignatureBytes()))
+	case 40:
+		messageClosingComplete := message.(*lnwire.ClosingComplete)
+		// FeeSatoshis should be u64 but it's i64 in LND
+		if messageClosingComplete.FeeSatoshis < 0 {
+			return nil
+		}
+		sb.WriteString("MSG_TYPE=closing_complete;CHANNEL_ID=")
+		sb.WriteString(fmt.Sprintf("%x", messageClosingComplete.ChannelID[:]))
+		sb.WriteString(";CLOSER_SCRIPTPUBKEY=")
+		sb.WriteString(fmt.Sprintf("%x", messageClosingComplete.CloserScript))
+		sb.WriteString(";CLOSEE_SCRIPTPUBKEY=")
+		sb.WriteString(fmt.Sprintf("%x", messageClosingComplete.CloseeScript))
+		sb.WriteString(";FEE_SATOSHIS=")
+		sb.WriteString(fmt.Sprintf("%d", messageClosingComplete.FeeSatoshis))
+		sb.WriteString(";LOCKTIME=")
+		sb.WriteString(fmt.Sprintf("%d", messageClosingComplete.LockTime))
+
+		extraBytes := 0
+
+		// LND actually doesn't check the signature when parsing the message,
+		// but we do, otherwise the fuzzer will crash all the time.
+		if messageClosingComplete.ClosingSigs.CloserAndClosee.IsSome() {
+			sig := messageClosingComplete.ClosingSigs.CloserAndClosee.UnsafeFromSome().Val
```

Comment:
Simplification here and below:

```suggestion
		messageClosingComplete.ClosingSigs.CloserAndClosee.WhenSomeV(func (sig Sig) {
```

Reply:
We can't actually do that, because of the early return inside the if statement.